README.md
# SwipeThroughText Flutter Package

[![Pub Version](https://img.shields.io/pub/v/swipe_through_text?style=plastic)](https://pub.dev/packages/swipe_through_text) [![GitHub](https://img.shields.io/github/license/PanaraStudios/swipe_through_text?style=plastic)](https://pub.dev/packages/swipe_through_text/license) 


SwipeThroughText is a customizable Flutter widget that allows users to swipe through a text and strikethrough it when a certain swipe threshold is met. The package is suitable for use in various types of Flutter apps, including to-do lists, notes, and more.


## Preview

<img src="https://raw.githubusercontent.com/PanaraStudios/swipe_through_text/master/preview/preview.gif" height="400"/>

## Limitations

> Currently this package only supports single line text, overflow text
> will be trimmed and ellipses `...` will be added the end. Still this
> is temporary limitation, feel free contribute to add support for
> multi-lines.

## Getting Started

### Installation

Add the following line to your `pubspec.yaml` file:

    dependencies:
      swipe_through_text: ^0.0.4 

Then, run `flutter pub get` in your terminal.

### Usage

Import the package into your Dart code:

    import 'package:swipe_through_text/swipe_through_text.dart';

To use the `SwipeThroughText` widget, simply add it to your widget tree with the required parameters, such as `text` and `textStyle`, and any optional parameters such as `strikethroughColor`, `strikethroughLineHeight`, `swipeThreshold`, `onSwipeComplete`, `onSwipeCancel`, and `dashArray`.

    SwipeThroughText(
      text: 'Swipe me!',
      textStyle: TextStyle(fontSize: 24),
      strikethroughColor: Colors.red,
      strikethroughLineHeight: 4,
      swipeThreshold: 0.8,
      onSwipeComplete: (fraction) {
        // Do something when swiping is complete
      },
      onSwipeCancel: (fraction) {
        // Do something when swiping is cancelled
      },
    )

You can also customize the strikethrough line using the `dashArray` parameter, which takes in a list of double values that determine the length of the dashes and gaps in the strikethrough line. For example, `[10, 5]` will create a dashed line with a 10-pixel dash and a 5-pixel gap.

### Example

    import 'package:flutter/material.dart';
    import 'package:swipe_through_text/swipe_through_text.dart';
    
    void main() {
      runApp(MyApp());
    }
    
    class MyApp extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          title: 'SwipeThroughText Demo',
          home: Scaffold(
            appBar: AppBar(
              title: Text('SwipeThroughText Demo'),
            ),
            body: Center(
              child: SwipeThroughText(
                text: 'Swipe me!',
                textStyle: TextStyle(fontSize: 24),
                strikethroughColor: Colors.red,
                strikethroughLineHeight: 4,
                swipeThreshold: 0.8,
                onSwipeComplete: (fraction) {
                  print('Swipe completed at $fraction');
                },
                onSwipeCancel: (fraction) {
                  print('Swipe cancelled at $fraction');
                },
                dashArray: [10, 5],
              ),
            ),
          ),
        );
      }
    }

## Contributing

Contributions to this package are welcome! If you find a bug or have a feature request, please open an issue or submit a pull request.


CHANGELOG.md
## 0.0.4

* Initial release of the SwipeThroughText Flutter package.
* Provides the SwipeThroughText widget for swiping through a text and strikethrough it when a certain swipe threshold is met.
* Provides several customizable parameters, such as textStyle, strikethroughColor, strikethroughLineHeight, and dashArray.
* Includes callbacks for onSwipeComplete and onSwipeCancel.
* Supports horizontal swiping and strikethrough animation.





lib/src/strike_through_text.dart
import 'package:flutter/material.dart';
import 'package:swipe_through_text/src/strike_through_painter.dart';

/// A widget that allows users to swipe through text and strikethrough it when a
/// certain threshold is met.
class SwipeThroughText extends StatefulWidget {
  /// The text to display.
  final String text;

  /// The style to use for the text.
  final TextStyle textStyle;

  /// The color to use for the strikethrough line.
  final Color strikethroughColor;

  /// The height to use for the strikethrough line.
  final double strikethroughLineHeight;

  /// The swipe threshold at which the strikethrough line is completed.
  final double swipeThreshold;

  /// The dash array to use for the strikethrough line.
  final List<double> dashArray;

  /// A callback that is called when the swipe is completed.
  final ValueChanged<double>? onSwipeComplete;

  /// A callback that is called when the swipe is cancelled.
  final ValueChanged<double>? onSwipeCancel;

  /// Creates a new [SwipeThroughText] instance.
  ///
  /// The [text], [textStyle], and [key] arguments are required.
  ///
  /// The [strikethroughColor], [strikethroughLineHeight], [swipeThreshold], and
  /// [dashArray] arguments have default values.
  ///
  /// The [swipeThreshold] value must be between 0.5 and 1.0 (inclusive).
  const SwipeThroughText({
    super.key,
    required this.text,
    required this.textStyle,
    this.strikethroughColor = Colors.red,
    this.strikethroughLineHeight = 2.0,
    this.swipeThreshold = 0.90,
    this.onSwipeComplete,
    this.onSwipeCancel,
    this.dashArray = const [],
  }) : assert(swipeThreshold >= 0.5 && swipeThreshold <= 1.0,
            'swipeThreshold must be between 0.5 - 1.0');

  @override
  SwipeThroughTextState createState() => SwipeThroughTextState();
}

/// The state for the [SwipeThroughText] widget.
class SwipeThroughTextState extends State<SwipeThroughText>
    with SingleTickerProviderStateMixin {
  double _strikethroughFraction = 0.0;
  bool _swipeRightToLeft = false;
  bool _swipeCompleted = false;
  double _initialSwipePosition = 0;
  double _textWidth = 0;

  late AnimationController _animationController;
  late Animation<double> _animation;

  @override
  void initState() {
    super.initState();

    // Create the animation controller and animation.
    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 200),
    );
    _animation = Tween<double>(begin: 0, end: 1).animate(_animationController)
      ..addListener(() {
        setState(() {
          _strikethroughFraction = _animation.value;
        });
      });
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  /// Measures the width of the text.
  ///
  /// The [maxWidth] argument is the maximum width available to the text.
  void _measureTextWidth(double maxWidth) {
    final textPainter = TextPainter(
      text: TextSpan(text: widget.text, style: widget.textStyle),
      textDirection: TextDirection.ltr,
      maxLines: 1,
      ellipsis: '...',
    )..layout(maxWidth: maxWidth);
    _textWidth = textPainter.width;
  }

  /// Handles the start of a swipe.
  ///
  /// The [localDx] argument is the local X coordinate of the swipe.
  ///
  /// The [maxWidth] argument is the maximum width available to the text.
  void _handleSwipeStart(double localDx, double maxWidth) {
    _initialSwipePosition = localDx;
    _swipeRightToLeft = localDx > maxWidth / 2;
  }

  /// Updates the strikethrough fraction based on the current swipe position.
  ///
  /// The [localDx] argument is the local X coordinate of the swipe.
  void _updateStrikethroughFraction(double localDx) {
    setState(() {
      double swipeDistance = _swipeRightToLeft
          ? _initialSwipePosition - localDx
          : localDx - _initialSwipePosition;
      _strikethroughFraction = (swipeDistance / _textWidth).clamp(0, 1);
    });
  }

  /// Handles the end of a swipe.
  void _handleSwipeEnd() {
    if (_strikethroughFraction >= widget.swipeThreshold) {
      _swipeCompleted = true;
      _strikethroughFraction = 1.0;
      setState(() {
        widget.onSwipeComplete?.call(_strikethroughFraction);
      });
    } else {
      widget.onSwipeCancel?.call(_strikethroughFraction);
      _animationController.reverse(from: _strikethroughFraction);
    }
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        _measureTextWidth(constraints.maxWidth);
        return GestureDetector(
          onHorizontalDragDown: _swipeCompleted
              ? null
              : (details) => _handleSwipeStart(
                  details.localPosition.dx, constraints.maxWidth),
          onHorizontalDragUpdate: _swipeCompleted
              ? null
              : (details) =>
                  _updateStrikethroughFraction(details.localPosition.dx),
          onHorizontalDragEnd:
              _swipeCompleted ? null : (details) => _handleSwipeEnd(),
          child: Stack(
            alignment: Alignment.center,
            children: [
              Text(
                widget.text,
                style: widget.textStyle,
                maxLines: 1,
                overflow: TextOverflow.ellipsis,
              ),
              Positioned(
                left: _swipeRightToLeft ? null : 0,
                right: _swipeRightToLeft ? 0 : null,
                child: CustomPaint(
                  size: Size(_textWidth * _strikethroughFraction,
                      widget.strikethroughLineHeight),
                  painter: StrikethroughPainter(
                    color: widget.strikethroughColor,
                    width: widget.strikethroughLineHeight,
                    dashArray: widget.dashArray,
                  ),
                ),
              ),
            ],
          ),
        );
      },
    );
  }
}


lib/src/strike_through_painter.dart
import 'package:flutter/material.dart';

/// A custom painter for drawing the strikethrough line.
class StrikethroughPainter extends CustomPainter {
  final Color color;
  final double width;
  final List<double> dashArray;

  /// Creates a new [_StrikethroughPainter] instance.
  ///
  /// The [color] and [width] arguments are required.
  ///
  /// The [dashArray] argument has a default value.
  StrikethroughPainter(
      {required this.color, required this.width, this.dashArray = const []});

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..strokeWidth = width
      ..style = PaintingStyle.stroke;

    if (dashArray.isNotEmpty) {
      final path = Path();
      double dashOffset = 0;
      bool draw = true;

      while (dashOffset < size.width) {
        double segmentLength = dashArray[dashOffset.toInt() % dashArray.length];
        if (draw) {
          path.moveTo(dashOffset, size.height / 2);
          path.lineTo(dashOffset + segmentLength, size.height / 2);
        }
        dashOffset += segmentLength;
        draw = !draw;
      }
      canvas.drawPath(path, paint);
    } else {
      canvas.drawLine(Offset(0, size.height / 2),
          Offset(size.width, size.height / 2), paint);
    }
  }

  @override
  bool shouldRepaint(covariant StrikethroughPainter oldDelegate) {
    return oldDelegate.color != color ||
        oldDelegate.width != width ||
        oldDelegate.dashArray != dashArray;
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:swipe_through_text/swipe_through_text.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'SwipeThroughText Demo',
      home: Scaffold(
        appBar: AppBar(
          title: const Text('SwipeThroughText Demo'),
        ),
        body: Container(
          alignment: Alignment.center,
          padding: const EdgeInsets.all(24),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SwipeThroughText(
                text: 'Swipe Through me! limited to 1 line',
                textStyle: const TextStyle(fontSize: 24),
                strikethroughColor: Colors.red,
                strikethroughLineHeight: 4,
                swipeThreshold: 0.5,
                onSwipeComplete: (fraction) {
                  debugPrint('Swipe completed at $fraction');
                },
                onSwipeCancel: (fraction) {
                  debugPrint('Swipe cancelled at $fraction');
                },
                dashArray: const [], // for dotted line use: const [10, 5],
              ),
              const SizedBox(
                height: 10,
              ),
              SwipeThroughText(
                text: 'Swipe Through me! limited to 1 line',
                textStyle: const TextStyle(fontSize: 24),
                strikethroughColor: Colors.green,
                strikethroughLineHeight: 4,
                swipeThreshold: 0.5,
                onSwipeComplete: (fraction) {
                  debugPrint('Swipe completed at $fraction');
                },
                onSwipeCancel: (fraction) {
                  debugPrint('Swipe cancelled at $fraction');
                },
                dashArray: const [10, 5],
              ),
              const SizedBox(
                height: 10,
              ),
              SwipeThroughText(
                text: 'Swipe Through me! limited to 1 line',
                textStyle: const TextStyle(fontSize: 24),
                strikethroughColor: Colors.purple,
                strikethroughLineHeight: 4,
                swipeThreshold: 0.5,
                onSwipeComplete: (fraction) {
                  debugPrint('Swipe completed at $fraction');
                },
                onSwipeCancel: (fraction) {
                  debugPrint('Swipe cancelled at $fraction');
                },
                dashArray: const [], // for dotted line use: const [10, 5],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
